### ISSUE TITLE
The ivector extractor in Kaldi-android is 30 times slower than Kaldi-linux

### ISSUE THREAD

**Issue by aoligeigeigei** [alphacep/vosk-api#544]
Hi, I'm currently working on move Kaldi to the Android. After I compiled the Kaldi-andriod and run my ivector case on Android, it's 30 times slower than Linux. 

The ivector running speed on Linux is 12ms, but on Android nearly 400ms. I know the speed will be slow if we move the Kaldi to Android, but that 30 times is a little out of whack. Based that, I tested the speed of xvector, 170ms on Linux and 700ms on Android. The speed of xvector exactly slower than ivector, but at least the slowing times is normal, This is where I'm really confused.

The main issue is ivector extractor after I checked, 99% of the time is spent on that. Does anybody have the same problem as me, I really need help to speed it up to 200ms.

NOTE: I closed all other programs on the Android phone when do the test.

**Comment by nshmyrev** [comment]
Android CPUs have very small cache thats why they have much bigger troubles multiplying matrices for i-vector extractor.

Our model uses smaller ivector dimension than standard kaldi recipe because of that (40 instead of 100).  You can try that too, it degrades accuracy a bit of course but much faster.

Alternatives would be to figure out how to access fast matrix computation on Android with NNAPI, but it is not really widely supported in hardware. Or you can try to quantize the models.

In general x-vectors and context-based models like contextnet have more chances to run faster on mobile devices.

**Comment by aoligeigeigei** [comment]
Hi @nshmyrev Thank you so much. I have solved my issue according the solution you mentioned. In addition to reducing the i-vector dimension, I adjusted the num_gauss and delta_order, and found the best model through different combinations of these three parameters. The speed was improved, and the accuracy didn't drop much. (The core issue is still reducing the model size, and all three parameters are related to the model size)
